### ISSUE TITLE
After re-initialising form, form submission will always succeed even if errors exist

### ISSUE THREAD

**Issue by TAGC** [erikras/redux-form#3456]
### Are you submitting a **bug report** or a **feature request**?

Bug report.

### What is the current behavior?

Given this code:

```js
function tryParseFloat (value: string): any {
  console.log('Parsing', value);
  return (value === '') ? value : parseFloat(value);
}

function validate (value: number | string): string | undefined {
  console.log('Validating', value, typeof(value));
  if (isUndefined(value)) return 'Required';
  if (value === '') return 'Required';
  return undefined;
}

function createPostFieldProperties (): PropertyDetails[] {
  const gainControl = (
      <Field
          name='gain'
          type='number'
          parse={tryParseFloat}
          validate={[validate]}
          component={renderField}
      />
  );
```

On successful submission, the form is reinialised:

```js
onSubmit: (values, _, props) => {
  const {form} = props;

  const onSuccess = () => {
    dispatch(destroy(definitionName)); // Invalidate the corresponding message definition form, if it exists.
    dispatch(initialize(form, values));
  };

  dispatch(isTx
      ? updateTxMessageDefinitionField(definitionName, fieldName, values, onSuccess)
      : updateRxMessageDefinitionField(definitionName, fieldName, values, onSuccess));
},
```

When first creating the form and trying to submit the form when the field does not contain a valid value (the value is undefined or an empty string), form submission fails as expected.

![2017-09-26_12-40-30](https://user-images.githubusercontent.com/1588951/30858571-266acde4-a2b8-11e7-805a-fd30d4ab97af.gif)

![electron_2017-09-26_12-47-23](https://user-images.githubusercontent.com/1588951/30858753-e3164e14-a2b8-11e7-8c14-d82edd3be872.png)

However, if I enter a valid value, submit the form (which succeeds as expected) and then try entering an invalid value, submission succeeds despite entering the invalid value. Redux still dispatches the correct `REDUX_SYNC_ERRORS` action with the expected payload, but `this.form` appears to be considered valid regardless, so submission succeeds:

![2017-09-26_12-43-49](https://user-images.githubusercontent.com/1588951/30858688-9a5c6050-a2b8-11e7-9c38-5c160ccbf3df.gif)

![electron_2017-09-26_12-48-58](https://user-images.githubusercontent.com/1588951/30859320-0caa27bc-a2bb-11e7-9044-878dcad72538.png)

### What is the expected behavior?

Invalid form data should continue to cause submission to fail even after a previous successful submission.

In other words, after I dispatch `initialize(form, values)` following a successful submission, I should still get `@@redux-form/SET_SUBMIT_FAILED` actions when validating for any fields in the form fails.

### What's your environment?

- Redux Form 6.7.0 (same issue occurs on 7.0.4 though)
- Windows 7 64-bit
- Electron 1.7.1

### Other information

I notice that this issue doesn't occur if I navigate elsewhere after making a successful submission (to cause the form to be destroyed and a new form to be created), and then re-navigate back and try submitting with invalid data.

The only difference I notice is that when navigating, a number of `@@redux-form/REGISTER_FIELD` actions get dispatched which doesn't occur when I simply dispatch the `initialize` action on successful submission.

**Comment by TAGC** [comment]
Made progress in tracking this down. The issue is indeed to do with the fact that my fields don't get re-registered after dispatching `initialize`. The absence of registered fields causes the `isValid(form, getFormState, false)(state);` to return `true` even if the form contains sync errors:

![image](https://user-images.githubusercontent.com/1588951/30968131-340ca6da-a456-11e7-93c2-0cce20342d2f.png)

What I need to figure out is how to reinitialise a form with the new initial values, while also re-registering fields (or preventing them from being de-registered in the first place

**Comment by TAGC** [comment]
I determined that fields are only unregistered (and therefore this issue only occurs) due to the fact that I'm destroying a *parent* form after successful submission:

```js
onSubmit: (values) => {
  const onSuccess = () => {
    dispatch(destroy(definitionName)); // Invalidate the corresponding message definition form, if it exists.
    // ...
  };
  // ...
```

To resolve this, I created a function to reinitialise my field definition forms, which gets called after a successful submission (as well as when the Redux component is first mounted):

```js
function reinitialiseFieldDefinitionForms () {
  VirtualFieldDefinitionForm = reduxForm({})(_VirtualFieldDefinitionForm);
  RollingFieldDefinitionForm = reduxForm({})(_RollingFieldDefinitionForm);
  ChecksumFieldDefinitionForm = reduxForm({})(_ChecksumFieldDefinitionForm);
}
```

```js
onSubmit: (values) => {
  const onSuccess = () => {
    dispatch(destroy(definitionName)); // Invalidate the corresponding message definition form, if it exists.
    reinitialiseFieldDefinitionForms();
  };
  // ...
```

This causes fields to be registered in addition to setting the initial values.
